### PR TITLE
Use a callback to create a HTTP client in Solr connector.

### PR DESCRIPTION
**Pull request by EreMaijala**
This makes it possible to create the client when the URL is available so that any proxy can be bypassed for local addresses.

This is provided as a starting point for solving the issue where proxy configuration prevents connection to a local Solr instance. If deemed appropriate, similar mechanism could be used to fix other cases where just passing HttpService in constructor instead of a Client is not sufficient.

TODO
- [x] Update changelog when merging

### CODE REVIEW

File: module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php [hunk 1]
```diff
@@ -102,19 +102,19 @@ class Connector implements \Laminas\Log\LoggerAwareInterface
...
         HandlerMap $map,
-        HttpClient $client,
+        callable $cf,
         $uniqueKey = 'id'
     ) {
         $this->url = $url;
         $this->map = $map;
         $this->uniqueKey = $uniqueKey;
-        $this->client = $client;
+        $this->clientFactory = $cf;
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php]: If we wanted to maintain backward compatibility, we could type sniff $cf, and if it's a Client object, define the factory as `function () use ($cf) { return clone $cf; }`. Not sure that there's a strong reason to do this (probably simpler not to), but just sharing a passing thought that might be useful if we want to prioritize backward-compatibility over simplicity.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php [hunk 2]
```diff
@@ -61,7 +62,17 @@ public function testCommit()
         );
         $connector = $bm->get('Solr')->getConnector();
         $connector->expects($this->once())->method('write')
-            ->with($this->isInstanceOf(CommitDocument::class));
+            ->with($this->isInstanceOf(CommitDocument::class))
+            ->will(
+                $this->returnCallback(
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: This seems like kind of a weird hack -- forcing a method to have a side effect it wouldn't normally have in order to replicate behavior that is being mocked out elsewhere (if I'm understanding correctly). Is there a more elegant way to do this?

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php [hunk 3]
```diff
@@ -582,6 +583,45 @@ public function testWriteDocument()
...
+     *
+     * @return void
+     */
+    public function testDelayedClientCreation()
+    {
+        $doc = new CommitDocument();
+        $client = $this->getMockBuilder(\Laminas\Http\Client::class)
+            ->onlyMethods(['setOptions'])
+            ->getMock();
+        $client->expects($this->exactly(0))->method('setOptions')
```
Comment (demiankatz) [inline module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php]: Why use `exactly(0)` rather than `never()`? If it's not going to be called, why are we specifying `with` parameters? Or am I misunderstanding the order of precedence here?

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php [hunk 4]
```diff
@@ -582,6 +583,45 @@ public function testWriteDocument()
...
+        $connector->expects($this->once())->method('write')
+            ->with(
+                $this->equalTo($doc),
+                $this->equalTo('update'),
+                $this->isNull()
+            )
+            ->will(
+                $this->returnCallback(
+                    function () use ($connector) {
+                        // Call client factory for expectations to be met:
```
Comment (demiankatz) [inline module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php:595]: Another instance of this hack.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php [hunk 5]
```diff
@@ -644,10 +684,19 @@ protected function loadResponse($fixture)
...
         $map = new HandlerMap(['select' => ['fallback' => true]]);
-        $client = $client ?? new \Laminas\Http\Client();
         return $this->getMockBuilder(\VuFindSearch\Backend\Solr\Connector::class)
             ->onlyMethods($mock)
-            ->setConstructorArgs(['http://example.org/', $map, $client])
+            ->setConstructorArgs(
+                [
+                    'http://localhost/',
+                    $map,
+                    function () use (&$client) {
```
Comment (demiankatz) [inline module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php]: Is there a reason you're passing a reference here, but not in WriterTest? Just curious about the apparent inconsistency.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/ConnectorTest.php [hunk 6]
```diff
@@ -161,7 +170,16 @@ public function testWriteJSON()
             ->with($this->equalTo($jsonData));
         $conn = $this->getMockBuilder(Connector::class)
```
Comment (demiankatz) [inline module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/ConnectorTest.php]: Is it worth having a `getMockConnector` function for this repeated code now that it's significantly longer?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, this makes sense to me in principle! I have a few questions/comments, mostly to do with the test suite. It seems to me that some of this might be able to be made more clear. If you want me to spend time coming up with a more tangible proposal for how, let me know. :-)

File: module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php]: That makes sense, added. There's no way to avoid changing the constructor signature, so it's a breaking change regardless in that way.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php (hunk 2 above)
Comment (EreMaijala) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: I removed the expectations that were not relevant to the functionality being tested in WriterTest.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php (hunk 3 above)
Comment (EreMaijala) [reply module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php]: That was a mistake. Tests have now been rearranged and cleaned up.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php (hunk 5 above)
Comment (EreMaijala) [reply module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/BackendTest.php]: Just a mistake.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/ConnectorTest.php (hunk 6 above)
Comment (EreMaijala) [reply module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/ConnectorTest.php]: Added, though named `getConnectorMock` for parity with the one in BackendTest.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php (hunk 2 above)
Comment (demiankatz) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: I think WriterTest is actually still testing more than it needs to, and we've lost the part of the test that confirms that expected timeout values are passed along. Since the actual behavior of the WriteDocumentCommand is covered by its own test, I don't think we need to create a whole environment to test that behavior here. Why mock a Solr Connector at all? As a unit test, we should just confirm that the Writer passes an appropriately populated WriteDocumentCommand to the search service... which I think we could do by just mocking the search service and putting a custom callback with expectations on the invoke method. What do you think? That would allow us to further simplify both testCommit and testOptimize. (It might also be worth rewriting other test methods, but it seems like commit/optimize is the bare minimum to keep things passing here, so I wouldn't ask for more than that).

Comment (EreMaijala) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: You're right of course. I've now pretty much rewritten all the tests. What do you think?

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php [hunk 7]
```diff
@@ -145,63 +120,64 @@ public function testDeleteAll()
...
+     * @param object $expectedCommand Expected command class
+     * @param mixed  $result          Result to return for the invoked command
+     *
+     * @return MockObject&SearchService
      */
-    protected function getMockChangeTracker()
+    protected function getMockSearchService($expectedCommand, $result)
     {
-        return $this->getMockBuilder(\VuFind\Db\Table\ChangeTracker::class)
+        $resultCommand = $this->getMockBuilder(WriteDocumentCommand::class)
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: Would it make sense for this to be `get_class($expectedCommand)` rather than a hard-coded value, or am I misunderstanding something?

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php [hunk 8]
```diff
@@ -45,22 +45,15 @@
...
      */
     public function testCommit()
     {
-        $bm = $this->getBackendManagerWithMockSolr();
-        $connector = $bm->get('Solr')->getConnector();
-        $connector->expects($this->once())->method('write')
-            ->with($this->isInstanceOf(CommitDocument::class));
-        $writer = new Writer($this->getSearchService($bm), $this->getMockChangeTracker());
-        $writer->commit('Solr');
+        $command = new WriteDocumentCommand('Solr', new CommitDocument(), 60 * 60);
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: Maybe it would make sense to rename all these $command variables to $expectedCommand, just to make it clearer what role they play in the test.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php [hunk 9]
```diff
@@ -145,63 +120,64 @@ public function testDeleteAll()
...
-                ->getMock();
-        $mockBackend->expects($this->any())->method('getConnector')
-            ->will($this->returnValue($mockConnector));
-        $mockBackend->expects($this->any())->method('getIdentifier')
-            ->will($this->returnValue('Solr'));
-        $sm->setService('Solr', $mockBackend);
-        return $pm;
+        return $this->getMockBuilder(\VuFind\Db\Table\ChangeTracker::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['markDeleted'])
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: I don't think this onlyMethods serves a purpose anymore, but I could be wrong.

**Review by demiankatz — COMMENTED** [review]
Thanks for the test revisions, I think this is a whole lot better now... just a few minor suggestions.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php (hunk 7 above)
Comment (EreMaijala) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: Makes sense, done. They all were instances of `WriteDocumentCommand`, but `get_class` makes the intention more obvious.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php (hunk 8 above)
Comment (EreMaijala) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: Done.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php (hunk 9 above)
Comment (EreMaijala) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Solr/WriterTest.php]: Right, removed.

**Review by demiankatz — APPROVED** [review]

**Comment by demiankatz** [comment]
@EreMaijala, this looks good to me now. I just have one remaining question: since it's no longer required to pass an HTTP factory function, should we revert test classes that were only changed to meet that requirement? It might make this overall PR more readable. No pressure, though!

I've approved the PR, and I'll leave the final decision to you whether you'd like to roll back a couple of tests or just merge as-is. Either way, feel free to go ahead as long as CI is passing.
